functional_programming_tests/test/with_functional_programming.cpp

```diff
@@ -0,0 +1,94 @@
+#include <gtest/gtest.h>
+
+#include <iostream>
+#include <vector>
+#include <cmath>
+#include <optional>
+
+// production_code.h/cc
+template<typename T>
+using Map = std::vector<std::vector<T>>;
+
+// X, Y position
+struct Pos {
+    size_t x;
+    size_t y;
+};
+
+using Path = std::vector<Pos>;
+
+// Operator overload to print path
+std::ostream& operator<<(std::ostream& os, std::vector<Pos> const& path) {
+    for (auto const& pose:path) {
+        os << "(" << pose.x << ", " << pose.y << ")\n";
+    }
+    return os;
+}
+
+// Operator overload for position comparison
+bool operator==(Pos const& lhs, Pos const& rhs) {
+    return lhs.x == rhs.x && lhs.y == rhs.y;
+}
+
+// From the Robot, goal and costmap, generate a trajectory (Deterministic calculation)
```

Comment:
I understand these comments are focused on being illustrative, but have you considered seeing what a doxygen version looks like?